File: GFG/Arrays/trapwater.py
# using precomputation to compute rather than computing for every index

def trap(arr):
    n = len(arr)
    wat=0
    rmaw=[]
    lmaw=[]
    lmaw.append(arr[0])
    rmaw.append(arr[n-1])
    i,j=1,n-2
    while(i<n and j>=0):
        lmaw.append(max(lmaw[i-1],arr[i]))
        rmaw.insert(0,max(rmaw[0],arr[j]))
        i+=1
        j-=1
    for i in range(n):
        wat+=min(lmaw[i],rmaw[i])-arr[i]
    return wat
        

brr =[9,5,4,3,2,10,2,3,4,5,7]

print(trap(brr))

File: GFG/Arrays/trapwater.cpp
#include<iostream>
#include<vector>
using namespace std;

int trap(vector<int> &arr){
    int n=arr.size();
    vector<int> lmax(n);
    vector<int> rmax(n);
    lmax[0]=arr[0];
    rmax[n-1]=arr[n-1];
    int i = 0,j=n-1;
    int wat=0;
    while(i<n && j>=0){
        lmax[i]=max(lmax[i-1],arr[i]);
        rmax[j]=max(rmax[j+1],arr[j]);
        i++;
        j--;
    }
    for(int i=0;i<n;i++){
        wat = wat+((min(lmax[i],rmax[i]))-arr[i]);
    }
    // for(int i=0;i<arr.size();i++){
    //     cout<<lmax[i]<<" "<<rmax[i]<<endl;
    // }
    return wat;
}

int main(){
    vector<int> arr={9,5,4,3,2,10,2,3,4,5,7};
    cout<<trap(arr);
    return 0;
}